app/main.py

```diff
@@ -1,6 +1,48 @@
-from typing import Callable
+from functools import wraps
+from typing import Callable, Dict, Any, Hashable
+
+
+def is_immutable(value: Any) -> bool:
+    """Check if a value is immutable."""
+    return isinstance(value, (int, float, str, tuple, frozenset, bool, type(None)))
+
+
+def make_cache_key(args: tuple, kwargs: dict) -> Hashable:
+    """Create a cache key that is a combination of args and frozenset for kwargs."""
+    return (args, frozenset(kwargs.items()))
```

Comment:
Using `frozenset` to create a cache key from `kwargs` may lead to issues if the order of keyword arguments changes between calls. Consider using a tuple of sorted items instead to ensure consistent ordering and avoid cache misses for semantically identical calls.